### PR TITLE
FCOPY-NEGATIVE.ps1: Refine the FCOPY-NEGATIVE test

### PR DESCRIPTION
**Pull request by xian123**
Main changes:
1. Replace check_fcopy_daemon() with Check-FcopyDaemon() in IntegrationServiceLibrary.psm1
2. Replace plink.exe with Run-LinuxCmd function to run Linux commands
3. Add a new step : verify the file can copy to vm when "Guest Service Interface" is enabled
4. Split the long lines into short ones

### CODE REVIEW

**Comment by xian123** [comment]
Test results with Ubuntu 18.04 and CentOS 7.5:
```
[LISAv2 Test Results Summary]
Test Run On           : 08/06/2019 01:45:11
VHD Under Test        : D:\xhx\vhds\Ubuntu_18.04.1.vhdx
Initial Kernel Version: 4.15.0-38-generic
Final Kernel Version  : 5.2.5-2519374d2a6b
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:10

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes) 
-------------------------------------------------------------------------------------------------------------------------------------------
    1 FCOPY                FCOPY-NEGATIVE                                                                    PASS                 6.43 
	

[LISAv2 Test Results Summary]
Test Run On           : 08/06/2019 02:13:51
VHD Under Test        : D:\xhx\vhds\Centos_7.5.vhdx
Initial Kernel Version: 3.10.0-862.el7.x86_64
Final Kernel Version  : 4.14.119-b0de27758059
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:6

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes) 
-------------------------------------------------------------------------------------------------------------------------------------------
    1 FCOPY                FCOPY-NEGATIVE                                                                    PASS                 2.16 
```

File: Testscripts/Windows/FCOPY-NEGATIVE.ps1 [hunk 1]
```diff
@@ -124,90 +124,104 @@ function Main {
...
-        $sts = Stop-FcopyDaemon -VmPassword $VMPassword -vmPort $VMPort -vmUserName $VMUserName -ipv4 $Ipv4
-        if (-not $sts[-1]) {
-            Write-LogErr "FAIL to stop hypervfcopyd inside the VM!"
-            return "FAIL"
-        }
+    #
+    # Step 4: verify the file cannot copy to vm when hypervfcopyd is stopped
+    #
+    Write-LogInfo "Info: Step 4: fcopy file to vm when hypervfcopyd stopped"
+    Enable-VMIntegrationService -Name "Guest Service Interface" -vmName $VMName -ComputerName $HvServer
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/FCOPY-NEGATIVE.ps1]: There is a chance the service name change. Please replace them with the global constant in the file. So we will update one place with new service name in future.

File: Testscripts/Windows/FCOPY-NEGATIVE.ps1 [hunk 2]
```diff
@@ -124,90 +124,104 @@ function Main {
...
-            Write-LogErr "File has been copied to guest VM even hypervfcopyd stopped"
-            return "FAIL"
-        }
-        elseif (($Error.Count -gt 0) -and ($Error[0].Exception.Message -like "*FAIL to initiate copying files to the guest*")) {
-            Write-LogInfo "Info: File could not be copied to VM as expected since hypervfcopyd stopped "
-        }
+    Copy-VMFile -vmName $VMName -ComputerName $HvServer -SourcePath $filePath `
+        -DestinationPath "/tmp/" -FileSource host -ErrorAction SilentlyContinue
+    if ($? -eq $true) {
+        Write-LogErr "File has been copied to guest VM even hypervfcopyd stopped"
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/FCOPY-NEGATIVE.ps1]: The File copied to

File: Testscripts/Windows/FCOPY-NEGATIVE.ps1 [hunk 3]
```diff
@@ -124,90 +124,104 @@ function Main {
...
     Write-LogInfo "Info: Step 1: fcopy file to vm when target folder is immutable"
 
-    # Verifying if /tmp folder on guest exists; if not, it will be created
-    .\Tools\plink.exe -C -pw $VMPassword -P $VMPort $VMUserName@$ipv4 "sudo [ -d /test ] || mkdir /test ; chattr +i /test"
-
+    # Verifying if /test folder on guest exists; if not, it will be created
+    $cmd = "[ -d /test ] || mkdir /test ; chattr +i /test"
+    Run-LinuxCmd -username $VMUserName -password $VMPassword -ip $Ipv4 -port $VMPort $cmd -runAsSudo
     if (-not $?) {
         Write-LogErr "Fail to change the permission for /test"
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/FCOPY-NEGATIVE.ps1:131]: I am asking if the line 131 and 132 are missing "return" statement.

File: Testscripts/Windows/FCOPY-NEGATIVE.ps1 [hunk 4]
```diff
@@ -124,90 +124,104 @@ function Main {
...
 
-    $Error.Clear()
-    Copy-VMFile -vmName $VMName -ComputerName $HvServer -SourcePath $filePath -DestinationPath "/test" -FileSource host -ErrorAction SilentlyContinue
-
-    if ( $? -eq $true ) {
-        Write-LogErr  "File has been copied to guest VM even  target folder immutable"
+    Copy-VMFile -vmName $VMName -ComputerName $HvServer -SourcePath $filePath `
+        -DestinationPath "/test" -FileSource host -ErrorAction SilentlyContinue
+    if ($? -eq $true) {
+        Write-LogErr  "File has been copied to guest VM even target folder immutable"
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/FCOPY-NEGATIVE.ps1]: The file copied to

File: Testscripts/Windows/FCOPY-NEGATIVE.ps1 [hunk 5]
```diff
@@ -124,90 +124,104 @@ function Main {
     #
     Write-LogInfo "Info: Step 1: fcopy file to vm when target folder is immutable"
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/FCOPY-NEGATIVE.ps1]: fcopy the file

File: Testscripts/Windows/FCOPY-NEGATIVE.ps1 [hunk 6]
```diff
@@ -124,90 +124,104 @@ function Main {
...
-        Write-LogInfo  "Info: File could not be copied to VM as expected since target folder immutable"
+
+    #
+    # Step 2: verify the file can copy to vm when "Guest Service Interface" is enabled
+    #
+
+    # Verifying if /tmp folder on guest exists; if not, it will be created
+    $cmd = "[ -d /tmp ] || mkdir /tmp"
+    Run-LinuxCmd -username $VMUserName -password $VMPassword -ip $ipv4 -port $VMPort $cmd -runAsSudo
+    Write-LogInfo "Info: Step 2: fcopy file to vm when 'Guest Service Interface' is enabled"
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/FCOPY-NEGATIVE.ps1]: fcopy the file to

File: Testscripts/Windows/FCOPY-NEGATIVE.ps1 [hunk 7]
```diff
@@ -124,90 +124,104 @@ function Main {
...
+    Copy-VMFile -vmName $VMName -ComputerName $HvServer -SourcePath $filePath `
+        -DestinationPath "/tmp/" -FileSource host -ErrorAction SilentlyContinue
+    if ($? -eq $true) {
+        Write-LogInfo "File has been copied to guest VM when 'Guest Service Interface' enabled"
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/FCOPY-NEGATIVE.ps1]: The file copied to

File: Testscripts/Windows/FCOPY-NEGATIVE.ps1 [hunk 8]
```diff
@@ -124,90 +124,104 @@ function Main {
...
     }
 
-    $Error.Clear()
-    Copy-VMFile -vmName $VMName -ComputerName $HvServer -SourcePath $filePath -DestinationPath "/tmp/" -FileSource host -ErrorAction SilentlyContinue
-
-    if ( $? -eq $true ) {
+    Copy-VMFile -vmName $VMName -ComputerName $HvServer -SourcePath $filePath `
+        -DestinationPath "/tmp/" -FileSource host -ErrorAction SilentlyContinue
+    if ($? -eq $true) {
         Write-LogErr "File has been copied to guest VM even 'Guest Service Interface' disabled"
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/FCOPY-NEGATIVE.ps1]: The file copied to

File: Testscripts/Windows/FCOPY-NEGATIVE.ps1 [hunk 9]
```diff
@@ -124,90 +124,104 @@ function Main {
...
-        #
-        # Step 3: verify the file cannot copy to vm when hypervfcopyd is stopped
-        #
-        Write-LogInfo "Info: Step 3: fcopy file to vm when hypervfcopyd stopped"
-        Enable-VMIntegrationService -Name "Guest Service Interface" -vmName $VMName -ComputerName $HvServer
-        if ( $? -ne $true) {
-            Write-LogErr "Fail to enable 'Guest Service Interface'"
-            return "FAIL"
-        }
+    Write-LogInfo "Info: File could not be copied to VM as expected since 'Guest Service Interface' disabled"
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/FCOPY-NEGATIVE.ps1]: The file

File: Testscripts/Windows/FCOPY-NEGATIVE.ps1 [hunk 10]
```diff
@@ -124,90 +124,104 @@ function Main {
...
-        # Stop fcopy daemon to do negative test
-        $sts = Stop-FcopyDaemon -VmPassword $VMPassword -vmPort $VMPort -vmUserName $VMUserName -ipv4 $Ipv4
-        if (-not $sts[-1]) {
-            Write-LogErr "FAIL to stop hypervfcopyd inside the VM!"
-            return "FAIL"
-        }
+    #
+    # Step 4: verify the file cannot copy to vm when hypervfcopyd is stopped
+    #
+    Write-LogInfo "Info: Step 4: fcopy file to vm when hypervfcopyd stopped"
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/FCOPY-NEGATIVE.ps1]: fcopy the file

File: Testscripts/Windows/FCOPY-NEGATIVE.ps1 [hunk 11]
```diff
@@ -124,90 +124,104 @@ function Main {
...
+        Write-LogErr "Fail to enable 'Guest Service Interface'"
+        return "FAIL"
+    }
 
-        $Error.Clear()
-        Copy-VMFile -vmName $VMName -ComputerName $HvServer -SourcePath $filePath -DestinationPath "/tmp/" -FileSource host -ErrorAction SilentlyContinue
+    # Stop fcopy daemon to do negative test
+    $sts = Stop-FcopyDaemon -VmPassword $VMPassword -vmPort $VMPort -vmUserName $VMUserName -ipv4 $Ipv4
+    if (-not $sts[-1]) {
+        Write-LogErr "FAIL to stop hypervfcopyd inside the VM!"
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/FCOPY-NEGATIVE.ps1]: Fail (Case consistency)

File: Testscripts/Windows/FCOPY-NEGATIVE.ps1 [hunk 12]
```diff
@@ -124,90 +124,104 @@ function Main {
...
+    # Verify the file does not exist after hypervfcopyd start
+    $cmd = "systemctl list-unit-files | grep fcopy"
+    $daemonName = Run-LinuxCmd -username $VMUserName -password $VMPassword -ip $Ipv4 -port $VMPort $cmd -runAsSudo
+    $daemonName = $daemonName.Split(".")[0]
+    Run-LinuxCmd -username $VMUserName -password $VMPassword -ip $Ipv4 -port $VMPort "systemctl start $daemonName" -runAsSudo
+    Start-Sleep -s 2
+    $sts = Check-FileInLinuxGuest -vmUserName $VMUserName -vmPassword $VMPassword `
+        -vmPort $VMPort -ipv4 $Ipv4 -fileName "/tmp/$testfile" -checkSize $true
+    if ($sts) {
+        Write-LogErr "File has been copied to guest vm after restart hypervfcopyd"
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/FCOPY-NEGATIVE.ps1]: The file copied to

File: Testscripts/Windows/FCOPY-NEGATIVE.ps1 [hunk 13]
```diff
@@ -124,90 +124,104 @@ function Main {
...
+        return "FAIL"
+    }
 
-        return "PASS"
+    # Removing the temporary test file
+    Remove-Item -Path \\$hvServer\$file_path_formatted -Force
+    if ($? -ne "True") {
+        Write-LogErr "Cannot remove the test file '${testfile}'!"
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/FCOPY-NEGATIVE.ps1]: Could not remove or Fail to remove

**Review by juhlee-microsoft — CHANGES_REQUESTED** [review]

File: Testscripts/Windows/FCOPY-NEGATIVE.ps1 (hunk 1 above)
Comment (xian123) [reply Testscripts/Windows/FCOPY-NEGATIVE.ps1]: Add a new global variable named VMIntegrationGuestService.

File: Testscripts/Windows/FCOPY-NEGATIVE.ps1 (hunk 3 above)
Comment (xian123) [reply Testscripts/Windows/FCOPY-NEGATIVE.ps1:131]: Add the 'return' statement.

**Comment by xian123** [comment]
Rerun the test case after updating the code and the result is PASS.

**Review by juhlee-microsoft — APPROVED** [review]

**Pull request merged**